rapid7/metasploit-payloads
# TLV traffic obfuscation with 4-non-null-byte XOR

This PR began as a small thought experiment after noticing just how much of Meterpeter's traffic was being analysed by various tools. AV solutions such as SONAR are getting better at picking up Meterpeter command strings that are going across the wire. SNORT rules are able to track the `RECV` in HTTP/S comms. Various other custom tools also look for the well-known command strings.

In short, there's lots of "signature" style detection of Meterpeter going on because of the content of the TLV packets. This PR aims to make that whole thing go away. It's based in part on various conversations with people, with sample implementations that are transport specific, and other ideas that have been rolling around in my head.

This PR requires changes on the MSF side as well. That PR can be found here: https://github.com/rapid7/metasploit-framework/pull/6480

## Design

These changes result in modification of the TLV communications such that for every packet (both sent and received):

* A 4-byte XOR key is generated. Each byte of this key is non-NULL.
* The TLV payload is encoded using this generated XOR, including the header.
* The XOR key is sent first, then the rest of the packet is sent after.
* When receiving, the 4 byte XOR key is read, then the next 8 bytes of the header. When decoded, the size of the packet is determined and the rest of the packet is read.
* When the entire packet has been read, the whole thing is then decoded.
* Once decoded, things carry on as they did before.

The aim here is to have a packet that looks different every time it's sent, even if the content of the packet is the same. Signature-based detections will have to level up.

## Status

This code isn't 100% complete. I've been dragging my feet a little getting the last bits done, but I wanted to get this PR going so that we could discuss it, and perhaps I could get some help from other people to finish it up.

### Stuff that works

* Windows x64 and x86 TCP, HTTP and HTTPS.
* Python TCP, HTTP and HTTPS
* PHP TCP
* Java TCP, HTTP and HTTPS
* Android TCP, HTTP and HTTPS

### Stuff that doesn't work (yet)

* POSIX x86 TCP -- haven't yet ported the code over from the Windows side, this is my next task. However, this is less of an issue, as AV/IDS tends not to impact POSIX. Still needs to be done though given MSF changes require it.

## Other stuff

Didier Stevens put together a snort rule that were able to detect Meterpreter sesions, and it was based on looking for `POST` requests with the `RECV` body. This PR also removes that, and instead we use `GET` requests without a body. This makes more sense HTTP wise as well.

## Testing

In my testing, I used HTTP for the most part because traffic analysis with wireshark is easy. All ya have to do is:

* Build and deploy a payload that works based on the above list (use HTTP to see traffic).
* Make sure you have the MSF PR running with the appropriate listener.
* Fire up wireshark.
* Run the payload.
* Watch the traffic in wireshark.

The packet contents should be completely unreadable and unrecognisable for each packet. Packets should never reuse the same key. You should no longer see the `RECV` calls at all.

## Help!

As I said, it's not finished yet. I'd appreciate any help from people who feel like getting stuck in to get this over the line.
